src/app/dashboard/homebrew/items/pack-detail/pack-detail.component.ts

```diff
@@ -145,6 +146,35 @@ export class PackDetailComponent implements OnInit, OnDestroy {
     });
   }
 
+  // validation
+  validatePackJSON(dialogRef, parsed) {
```

Comment:
This function could be split up into 2 separate bit of logic, which would let you remove a mutable variable (`valid`) and move the actual validation logic into a pure, easily testable function.

```
isValid(parsed) {
  if (parsed instanceof Array) {
    return Boolean(parsed.length) && parsed.every(item => this.objectIsItem(item));
  } else if (parsed) {
    return this.objectIsItem(parsed);
  }
  return false;
}

validatePackJSON(dialogRef, parsed) {
  dialogRef.componentInstance.loading = false;
  if (isValid(parsed)) {
    dialogRef.close(JSON.parse(dialogRef.componentInstance.data));
  } else {
    dialogRef.componentInstance.error = 'Invalid pack data';
  }
}
```